sphinx/source/docs/user_guide/examples/extensions_example_tool.py

```diff
@@ -2,51 +2,63 @@
 from bokeh.io import output_file, show
 from bokeh.models import ColumnDataSource, Tool
 from bokeh.plotting import figure
+from bokeh.util.compiler import TypeScript
 
 output_file('tool.html')
 
-JS_CODE = """
+TS_CODE = """
 import * as p from "core/properties"
 import {GestureTool, GestureToolView} from "models/tools/gestures/gesture_tool"
+import { GestureEvent } from "core/ui_events"
+import { ColumnDataSource } from "models/sources/column_data_source"
 
 export class DrawToolView extends GestureToolView
-
-  # this is executed when the pan/drag event starts
-  _pan_start: (e) ->
-    @model.source.data = {x: [], y: []}
-
-  # this is executed on subsequent mouse/touch moves
-  _pan: (e) ->
-    frame = @plot_model.frame
-
-    {sx, sy} = e
-    if not frame.bbox.contains(sx, sy)
-      return null
-
-    x = frame.xscales['default'].invert(sx)
-    y = frame.yscales['default'].invert(sy)
-
-    @model.source.data.x.push(x)
-    @model.source.data.y.push(y)
-    @model.source.change.emit()
-
-  # this is executed then the pan/drag ends
-  _pan_end: (e) -> return null
-
-export class DrawTool extends GestureTool
-  default_view: DrawToolView
-  type: "DrawTool"
-
-  tool_name: "Drag Span"
-  icon: "bk-tool-icon-lasso-select"
-  event_type: "pan"
-  default_order: 12
-
-  @define { source: [ p.Instance ] }
+{
+  model:DrawTool
+
+  //this is executed when the pan/drag event starts
+  _pan_start(_ev: GestureEvent): void{
+    this.model.source.data = {x: [], y: []}
+  }
+
+  //this is executed on subsequent mouse/touch moves
+  _pan(ev: GestureEvent): void{
+    const frame = this.plot_model.frame
+    const {sx, sy} = ev
+    if (!frame.bbox.contains(sx,sy)){return}
+    const x = frame.xscales['default'].invert(sx)
+    const y = frame.yscales['default'].invert(sy)
+
+    this.model.source.get_array('x').push(x)
+    this.model.source.get_array('y').push(y)
+    this.model.source.change.emit()
+  }
+
+  // this is executed then the pan/drag ends
+  _pan_end(_ev: GestureEvent):void{}
+}
+
+export class DrawTool extends GestureTool {
+  source: ColumnDataSource
+  default_view = DrawToolView
+
+  tool_name= "Drag Span"
+  icon= "bk-tool-icon-lasso-select"
+  event_type= "pan" as "pan"
+  default_order= 12
+
+  static initClass(): void {
+    this.define({
+    source: [p.Any],
+    })
+  }
+  type = "DrawTool"
```

Comment:
`type` and `default_view` should be defined on the `protocol`, e.g.:
https://github.com/bokeh/bokeh/blob/d260c429451a04b1fc00f593ef98f10ce1392e57/bokehjs/src/lib/models/annotations/label.ts#L118-L120

Reply:
I don't really understand what the purpose of the prototype.
Do you have some references which explain this?
Same goes for `Attrs` and `Props`  I don't understand the aim.